firedrake/supermeshing.py

```diff
@@ -0,0 +1,432 @@
+        "num_nodes_A": num_nodes_A,
+        "num_nodes_B": num_nodes_B,
+        "value_size_A": V_A.value_size,
+        "value_size_B": V_B.value_size,
+        "libsupermesh_simplex_measure": "libsupermesh_triangle_area" if dim == 2 else "libsupermesh_tetrahedron_volume",
+        "libsupermesh_intersect_simplices": "libsupermesh_intersect_tris_real" if dim == 2 else "libsupermesh_intersect_tets_real",
+        "dim": dim
+    }
+
+    dirs = get_petsc_dir() + (os.environ["VIRTUAL_ENV"], )
+    includes = ["-I%s/include" % d for d in dirs]
+    libs = ["-L%s/lib" % d for d in dirs]
+    libs = libs + ["-Wl,-rpath,%s/lib" % d for d in dirs] + ["-lpetsc", "-lsupermesh"]
+    lib = load(supermesh_kernel_str, "c", "supermesh_kernel",
+               cppargs=includes,
+               ldargs=libs,
+               argtypes=[ctypes.c_voidp, ctypes.c_voidp, ctypes.c_voidp, ctypes.c_voidp, ctypes.c_voidp, ctypes.c_voidp, ctypes.c_voidp],
+               restype=ctypes.c_int)
+
+    ammm(V_A, V_B, likely, node_locations_A, node_locations_B, M_SS, ctypes.addressof(lib), mat)
+    # for cell_A in range(len(V_A.cell_node_map().values)):
+    #     for cell_B in likely(cell_A):
+    #         simplex_A = vertices_A[vertex_map_A[cell_A, :], :].flatten()
+    #         simplex_B = vertices_B[vertex_map_B[cell_B, :], :].flatten()
+    #         lib(simplex_A.ctypes.data, simplex_B.ctypes.data, simplices_C.ctypes.data,
+    #             node_locations_A.ctypes.data, node_locations_B.ctypes.data,
+    #             M_SS.ctypes.data, outmat.ctypes.data)
+    #         print("outmat:\n", outmat)
+    #         print("dofs_A:\n", V_A.cell_node_list[cell_A])
+    #         print("dofs_B:\n", V_B.cell_node_list[cell_B])
+    #         mat.setValuesLocal(V_B.cell_node_list[cell_B], V_A.cell_node_list[cell_A], outmat, addv=PETSc.InsertMode.ADD_VALUES)
+
+    # Compute M_AB:
+    # For cell_A in mesh_A:
+    #     For cell_B in likely(cell_A):
+    #         mesh_S = supermesh(cell_A, cell_B)
+    #         if mesh_S is empty: continue
+    #         For cell_S in mesh_S:
+    #             evaluate basis functions of cell_A at dofs(A) of cell_S -> R_AS matrix
+    #             scale precomputed mass matrix to get M_SS
+    #                   (or mixed mass matrix if V_A, V_B have different finite elements)
+    #             evaluate basis functions of cell_B at dofs(B) of cell_S -> R_BS matrix
+    #             compute out = R_BS^T @ M_SS @ R_AS with dense matrix triple product
+    #             stuff out into relevant part of M_AB (given by outer(dofs_B, dofs_A))
+
+    mat.assemble()
+
+    if orig_value_size == 1:
+        return mat
+    else:
+        (lrows, grows), (lcols, gcols) = mat.getSizes()
+        lrows *= orig_value_size
+        grows *= orig_value_size
+        lcols *= orig_value_size
+        gcols *= orig_value_size
+        size = ((lrows, grows), (lcols, gcols))
+        context = BlockMatrix(mat, orig_value_size)
```

Comment:
Make a python matrix that is `mass \otimes I` where `I` is an appropriate identity matrix.